### PR TITLE
Use Gulp to build the client

### LINKED ISSUE #21: Use webpack for building and minifying.
This project really needs a proper bundler. Will see if any of my existing webpack template files will work.

We should also switch to yarn as a package manager.

2019-09-30:
We need to build the actual application bundle as well as just the dependency bundle and minify.

### LINKED ISSUE #546: Web interface broken in some browsers like Waterfox or IceWeasel
When Using the Web interface on a desktop browser the listed tabs in the dashboard fail to load
General
Playback
DLNA
Live TV
DVR
Scheduled Tasks
any Plugin added tab
and The Metadata Page
However This doesnt happen at all on mobile browsers and the android app
Tested desktop Browsers include Firefox,waterfox,opera,Microsoft edge, Internet explorer
Tested Mobile browsers are Icecat, Chrome, Unamed sony made browser, and the official android app(need to test tv one)
when trying on a fresh install the wizard is affected as well.

The background is on a Win10 x64 machine the tested ones that replicate it are the portable install of 4.0 and the installer. Originally a 3.7 portable was on there i have tested a clean install as well with both version of 4.0 no profiles with and without installed as service complete removal of appdata folder and the like still causes this. I have tested with adblockers and script blockers off.

**Logs**
[log_20191020.log](https://github.com/jellyfin/jellyfin/files/3748034/log_20191020.log)

**Screenshots**
jellyfin issue
as shown the browser navigates to the general tab but nothing loads the working tabs work as intended. 
jellyfin issue 2
This is the wizard the box is expanded to select language.

 - OS: Win10x64 portable and install
 - Browser: Waterfox 56.2.12 and 56.2.14, Microsoft Edge, Icecat mobile and chrome mobile Noting that requesting the desktop site there doesnt break the tabs
 - Jellyfin Version:10.4.0
 - Reverse proxy: none

### PR DESCRIPTION
**Pull request by MrTimscampi**
**Changes**

Turns out it's a lot easier to build the client properly when using a task runner than when using a bundler.

As of aea1740, this does the following:

- [x] Clean the `dist` folder
- [x] Handle .js files
  - [x] Run Babel (transpiling)
  - [x] Run Terser (minifying)
  - [x] Generate sourcemaps
- [x] Handle bundle.js
  - [x] Run webpack
  - [x] Handle production and development configs
  - [x] Run Babel on JS (transpiling)
  - [x] Run autoprefixer on CSS
  - [x] Copy assets
- [x] Handle .css files
  - [x] Compile SASS to CSS
  - [x] Run cssnano (minifying)
  - [x] Run autoprefixer
  - [x] Generate sourcemaps
- [x] Handle .html files
  - [x] Inject appLoader.js in index.html
  - [x] Run html-minifier
  - [x] Handle standalone mode
- [x] Support development and production environments
  - [x] Include sourcemaps only in development environments
- [x] Provide a dev server
- [x] Support live reloading

Build time is currently around 21s on my machine.

**Issues**

Helps with #836, fixes #21, probably fixes #546

### CODE REVIEW

**Comment by thornbill** [comment]
Would now be a good time to add sass support?

**Comment by MrTimscampi** [comment]
> Would now be a good time to add sass support?

Yes, it's planned as part of the css task :)

File: gulpfile.js [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+const inject = require('gulp-inject');
+const postcss = require('gulp-postcss');
+const cssnano = require('cssnano');
+const autoprefixer = require('autoprefixer');
+const sass = require('gulp-sass');
+ 
+sass.compiler = require('node-sass')
+
+
+if(mode.production()) {
```
Comment (JustAMan) [inline gulpfile.js]:
```suggestion
if (mode.production()) {
```

File: gulpfile.js [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+    .pipe(sass().on('error', sass.logError))
+    .pipe(postcss())
+    .pipe(mode.development(sourcemaps.write('.')))
+    .pipe(dest('dist/'))
+    .pipe(browserSync.stream());
+}
+
+function html() {
+    return src(['src/**/*.html', '!src/index.html'], {base: './src/'})
+    .pipe(htmlmin({ collapseWhitespace: true }))
```
Comment (JustAMan) [inline gulpfile.js]: can this be turned off in development mode?

File: src/scripts/site.js [hunk 3]
```diff
@@ -721,7 +721,6 @@ var AppInfo = {};
             },
             bundles: {
                 bundle: [
-                    "document-register-element",
```
Comment (JustAMan) [inline src/scripts/site.js]: why?

Comment (MrTimscampi) [reply src/scripts/site.js]: Mistake from my core-js trial :p Good catch

**Comment by Narfinger** [comment]
Please note that the Dockerfile in jellyfin/jellyfin needs to be changed a bit as some packages need to be installed. Notice that this might be a superset of things that need to be installed.
RUN apk add curl git zlib zlib-dev autoconf g++ make libpng-dev gifsicle alpine-sdk automake libtool make gcc musl-dev nasm

**Comment by EraYaN** [comment]
Wait why does it need g++?

**Comment by Narfinger** [comment]
> Wait why does it need g++? I suppose only in the `web-builder` container?
The current config needs gulp-imagemin which needs mozjpeg and gifsicle which on my docker system does compile some stuff. I am myself not sure perhaps the following is enough (according to some other thread):
 apk add alpine-sdk autoconf automake bash build-base git lcms2-dev libjpeg-turbo libjpeg-turbo-dev libpng libpng-dev libwebp libwebp-dev nasm zlib zlib-dev

**Comment by MrTimscampi** [comment]
> The current config needs gulp-imagemin which needs mozjpeg and gifsicle 

That's pretty much it, yeah. gulp-imagemin uses imagemin, which uses mozjpeg and gifsicle to process JPEGs and GIFs respectively.  
NPM builds the dependencies, hence the need for the new dependencies.

File: gulpfile.js [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+            ['@babel/preset-env']
+        ]
+    }))
+    .pipe(terser({
+        keep_fnames: true,
+        mangle: false
+    }))
+    .pipe(mode.development(sourcemaps.write('.')))
+    .pipe(dest('dist/'))
+    .pipe(browserSync.stream());;
```
Comment (dkanada) [inline gulpfile.js]:
```suggestion
    .pipe(browserSync.stream());
```

File: gulpfile.js [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+}
+
+function css() {
+    return src(['src/**/*.css', 'src/**/*.scss'], {base: './src/'})
+    .pipe(mode.development(sourcemaps.init({loadMaps: true})))
+    .pipe(sass().on('error', sass.logError))
+    .pipe(postcss())
+    .pipe(mode.development(sourcemaps.write('.')))
+    .pipe(dest('dist/'))
+    .pipe(browserSync.stream());
```
Comment (dkanada) [inline gulpfile.js]: I would indent all this chaining.

File: gulpfile.js [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+const concat = require('gulp-concat');
+const terser = require('gulp-terser');
+const htmlmin = require('gulp-htmlmin');
+const imagemin = require('gulp-imagemin');
+const sourcemaps = require('gulp-sourcemaps');
+const mode = require('gulp-mode')({
+    modes: ["production", "development"],
+    default: "development",
+    verbose: false
+  });
```
Comment (dkanada) [inline gulpfile.js]:
```suggestion
});
```

File: postcss.config.js [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+const cssnano = require('cssnano');
+
+const config = () => ({
+  plugins: [
+    postcssPresetEnv(),
+    cssnano()
+  ]
+})
+
+module.exports = config
```
Comment (dkanada) [inline postcss.config.js]: newline

**Comment by MrTimscampi** [comment]
@dkanada Fixed the few points in you review and added core-js polyfills via @babel/polyfills :)

**Comment by MrTimscampi** [comment]
What the hell happened here...

File: postcss.config.js [hunk 8]
```diff
@@ -0,0 +1,11 @@
+const postcssPresetEnv = require('postcss-preset-env')
```
Comment (JustAMan) [inline postcss.config.js]:
```suggestion
const postcssPresetEnv = require('postcss-preset-env');
```

File: src/polyfill.js [hunk 9]
```diff
@@ -0,0 +1,7203 @@
+(function(){function r(e,n,t){function o(i,f){if(!n[i]){if(!e[i]){var c="function"==typeof require&&require;if(!f&&c)return c(i,!0);if(u)return u(i,!0);var a=new Error("Cannot find module '"+i+"'");throw a.code="MODULE_NOT_FOUND",a}var p=n[i]={exports:{}};e[i][0].call(p.exports,function(r){var n=e[i][1][r];return o(n||r)},p,p.exports,r,e,n,t)}return n[i].exports}for(var u="function"==typeof require&&require,i=0;i<t.length;i++)o(t[i]);return o}return r})()({1:[function(_dereq_,module,exports){
```
Comment (JustAMan) [inline src/polyfill.js]: what is this and why it is checked in?

Comment (MrTimscampi) [reply src/polyfill.js]: Moved to bundle.js (Not sure if it actually works this way, but it should)

**Pull request merged**